pennylane_qulacs/qulacs_device.py

```diff
@@ -39,181 +74,240 @@
-                gate_matrix = mapped_operation(*par)
-            else:
-                gate_matrix = mapped_operation
+            self._state = QuantumState(self.num_wires)
 
-            dense_gate = gate.DenseMatrix(wires, gate_matrix)
-            self._circuit.add_gate(dense_gate)
-        else:
-            mapped_operation = self._operations_map[operation]
-            self._circuit.add_gate(mapped_operation(*wires, *par))
+        self._circuit = QuantumCircuit(self.num_wires)
 
-    @property
-    def state(self):
-        return self._state.get_vector()
+    def apply(self, operations, **kwargs):
+        rotations = kwargs.get("rotations", [])
 
-    def pre_measure(self):
-        self._circuit.update_quantum_state(self._state)
+        self.apply_operations(operations)
 
-    def expval(self, observable, wires, par):
-        bra = self._state.copy()
+        # Rotating the state for measurement in the computational basis
+        self.apply_operations(rotations)
 
-        if isinstance(observable, list):
-            A = self._get_tensor_operator_matrix(observable, par)
-            wires = [item for sublist in wires for item in sublist]
-        else:
-            A = self._get_operator_matrix(observable, par)
+    def apply_operations(self, operations):
+        """Apply the circuit operations to the state.
 
-        dense_gate = gate.DenseMatrix(wires, A)
-        dense_gate.update_quantum_state(self._state)
+        This method serves as an auxiliary method to :meth:`~.QulacsDevice.apply`.
 
-        expectation = inner_product(bra, self._state)
+        Args:
+            operations (List[pennylane.Operation]): operations to be applied
+        """
 
-        return expectation.real
+        for op in operations:
+            wires = op.wires
+            par = op.parameters
 
-    def probabilities(self):
-        states = itertools.product(range(2), repeat=self.num_wires)
-        probs = np.abs(self.state)**2
+            mapped_operation = self._operation_map[op.name]
+            if op.name.endswith(".inv"):
+                mapped_operation = self._get_inverse_operation(mapped_operation, wires, par)
 
-        return OrderedDict(zip(states, probs))
+            if op.name in ["QubitStateVector", "QubitStateVector.inv"]:
+                input_state = par[0]
+                input_state = _reverse_state(input_state)
 
-    def reset(self):
-        self._state.set_zero_state()
-        self._circuit = QuantumCircuit(self.num_wires)
+                if len(input_state) != 2**len(wires):
+                    raise ValueError("State vector must be of length 2**wires.")
+                if not np.isclose(np.linalg.norm(input_state, 2), 1.0, atol=tolerance):
+                    raise ValueError("Sum of amplitudes-squared does not equal one.")
+                # call qulac"s state initialization
+                self._state.load(input_state)
+
+            elif op.name in ["BasisState", "BasisState.inv"]:
```

Comment:
Could include if these operations are the first in the circuit

Reply:
Oh I assumed the qnode would check it. Reinserted!

Reply:
Plus test.